File: go/202_Happy_Number/202_Happy_Number.go
func isHappy(n int) bool {
    m := make(map[int]bool)
    for n > 1 {
        if _, exists := m[n]; exists {
            return false
        }
        m[n] = true
        n = sumNum(n)
    }
    return true
}

func sumNum(n int) int {
    total := 0
    for n > 0 {
        digit := n % 10
        n = n / 10
        total += digit * digit
    }
    return total
}

File: java/202_Happy_Number/202_Happy_Number.java
class Solution {
    public boolean isHappy(int n) {
        HashMap<Integer, Boolean> map = new HashMap<Integer, Boolean>();
        while (n > 1) {
            if (map.containsKey(n)) {
                return false;
            }
            map.put(n, true);
            n = sumNum(n);
        }
        return true;
    }
    
    private int sumNum(int n) {
        int total = 0;
        while (n > 0) {
            int digit = n % 10;
            n = n / 10;
            total += digit * digit;
        }
        return total;
    }
}
